### DISCUSSION TITLE
Add warning to prevent async lambdas in sync delegates

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5578] [Language Ideas]
```cs
public delegate void SyncDelegate();
SyncDelegate syncDel = async () => await System.Threading.Tasks.Task.Delay(100);
```
The above compiles without any issues, however I think it should be an error. The "natural" type for an `async` lambda should return `Task`, not `void` (which is what I think happens here).
Given that this is already out in the wild, I think adding an error would be too much of a breaking change, but a warning would definitely work - it would bring the developers' attention to the issue.
I'm writing this because I just found a bug this has caused in my code - this is a real problem.

**Comment by huoyaoyuan** [comment]
The pattern (async lambda in sync delegate) is heavily used in UI code. For example:

```csharp
button.Click += async (sender, e) =>
{
    button.IsEnabled = false;
    var result = await GetResultAsync();
    textBlock.Text = result.ToString();
    button.IsEnabled = true;
}
```

Yes, `async void` can cause problem. There can be an opt-in rule to warn on all `async void`.

**Reply by TahirAhmadov** [reply]
I hadn't thought about this, actually. Yes I can imagine this is heavily used. Perhaps we can make it a warning, and create a new syntax `async void (sender, e) => { ... }`? Then IDE can warn and suggest a correction.

**Comment by HaloFour** [comment]
Sounds like a more appropriate use case for an analyzer.  An analyzer would also have the benefit of being able to be more sophisticated and opinionated regarding how it determines use of the feature to be "inappropriate".

**Comment by CyrusNajmabadi** [comment]
> The "natural" type for an async lambda should return Task

There's no natural type here. The natural type comes into play of there is no contextual type. However, as there is a compatible contextual type, that is what the type of the lambda is. This is how lambdas have worked since c#3

**Reply by TahirAhmadov** [reply]
I used the wrong term, but I'm not sure how to word it correctly. 
In any case, would you be receptive to what I suggested above - making it a warning, and asking developers to change it to `async void (sender, e) => { ... }` with an IDE auto-fix?

**Reply by CyrusNajmabadi** [reply]
No. We would not warn here. You could request an analyzer at roslyn-analyzers. 

Independent of that, it would be fine to have an ide refactoring to add the return type to the lambda.
